lib/Transforms/HandshakeOptimization.cpp

```diff
@@ -0,0 +1,629 @@
+#include "dynamatic/Transforms/HandshakeOptimization.h"
+#include "dynamatic/Dialect/Handshake/HandshakeCanonicalize.h"
+#include "dynamatic/Dialect/Handshake/HandshakeOps.h"
+#include "dynamatic/Support/LLVM.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Value.h"
+#include "mlir/IR/ValueRange.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/raw_ostream.h"
+#include <vector>
+using namespace mlir;
+using namespace dynamatic;
+
+namespace {
+
+struct RemoveBranchMUXPairs
+    : public OpRewritePattern<handshake::ConditionalBranchOp> {
+  using OpRewritePattern<handshake::ConditionalBranchOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(handshake::ConditionalBranchOp condBranchOp,
+                                PatternRewriter &rewriter) const override {
+    mlir::Operation *useOwner1 = nullptr;
+    mlir::Operation *useOwner2 = nullptr;
+
+    auto it = condBranchOp.getTrueResult().getUsers().begin();
```

Comment:
The following avoid the double verbose call to `condBranchOp.getTrueResult().getUsers()` and makes it clear that if any of the branch's results has no user then the pattern cannot apply.

```cpp
auto trueResUsers = condBranchOp.getTrueResult().getUsers();
if (trueResUsers.empty())
  // This should not even be possible if we assume the IR
  // is materialized, so we should not even have to check for it, but
  // since we are super unclear on what the pass requirements actually
  // are I am leaving it for now
  return failure()
userOwner1 = *trueResUsers.begin();
```